optuna/study/_optimize.py

```diff
@@ -237,93 +222,38 @@ def _run_trial(
 
     # `Study.tell` may raise during trial post-processing.
     try:
-        study.tell(trial, values=values, state=state)
+        with warnings.catch_warnings(record=True) as tell_warnings:
+            warnings.simplefilter("always", UserWarning)
+            frozen_trial = study.tell(trial, values=value_or_values, state=state)
     except Exception:
+        frozen_trial = study._storage.get_trial(trial._trial_id)
         raise
     finally:
-        if state == TrialState.COMPLETE:
-            study._log_completed_trial(trial, cast(List[float], values))
-        elif state == TrialState.PRUNED:
-            _logger.info("Trial {} pruned. {}".format(trial.number, str(func_err)))
-        elif state == TrialState.FAIL:
+        if frozen_trial.state == TrialState.COMPLETE:
+            study._log_completed_trial(frozen_trial)
+        elif frozen_trial.state == TrialState.PRUNED:
+            _logger.info("Trial {} pruned. {}".format(frozen_trial.number, str(func_err)))
+        elif frozen_trial.state == TrialState.FAIL:
             if func_err is not None:
                 _logger.warning(
                     "Trial {} failed because of the following error: {}".format(
-                        trial.number, repr(func_err)
+                        frozen_trial.number, repr(func_err)
                     ),
                     exc_info=func_err_fail_exc_info,
                 )
-            elif values_conversion_failure_message is not None:
-                _logger.warning(values_conversion_failure_message)
-            else:
-                assert False, "Should not reach."
+            if tell_warnings is not None:
```

Comment:
Are these two `if` statements ever both evaluated to true? I'm curious since the first one is generally when the objective value raises, and the latter if tell fails due to e.g. bad return values from the objective.

Reply:
As you pointed out, these rules should not be evaluated to be `True`: the first happens when an exception is raised in `func()`, which makes `state` to be `PRUNED` or `FAIL` and `value_or_values` to be `None`. For the case, the latter is always invoked with `state` not to be `None` (and `skip_if_finished` to be `False`).

Do you think these two if rules should be unified to one if-elif block?

Reply:
> Do you think these two if rules should be unified to one if-elif block?

Not a strong opinion but preferably yes. One reasons is, the latter statement doesn't have to be evaluated, if the first one is true. It'll also improve code readability in my opinion.

Reply:
👍. 884bdd3 unify them.